File: euler_088.py
from time import time
from numpy import prod

def is_ps(t):
    return sum(t) == prod(t)

def product_sum_tuple(t):
    p = prod(t)
    s = sum(t)
    k = p - s
    l = len(t) + k

    return (l, p)


start = time()

# Non-trivial divisor of all numbers up to 24000
divisors = {n:[] for n in range(2, 24001)}

for d in range(2, 12001):
    k = 2 * d
    while k < 24001:
        divisors[k].append(d)
        k += d

# All non-trivial decompositions of n
def decomp(n):
    divs = divisors[n]

    if len(divs) == 0:
        return []

    if len(divs) == 1:
        return [(divs[0], n // divs[0])]

    comps = []

    for d in divs:
        comps.append(tuple(sorted((d, n // d))))
        for t in decomp(n // d):
            comps.append(tuple(sorted([d] + sorted(t))))

    return set(comps)


factorizations = {n:decomp(n) for n in range(2, 24001)}

# All set sizes given a product sum number n
set_sizes = {n:set() for n in range(2, 24001)}

for n in range(2, 24001):
    for f in factorizations[n]:
        set_sizes[n].add(product_sum_tuple(f)[0])

# All product sum numbers given a set size
psn = {k:set() for k in range(2, 12001)}

for n in range(2, 24001):
    for s in set_sizes[n]:
        if s < 12001:
            psn[s].add(n)


# Unique minimal product sum numbers
unique_minimal = set()

for s in psn.values():
    if s:
        unique_minimal.add(min(s))


print("Sum of minimal product sum numbers: ", sum(unique_minimal))

stop = time()
print("Time: " + str(stop - start))

File: euler_088.js
const start = Date.now();

// The minimal product-sum number with "size" k has to satisfy
// k <= M <= 2k, so 2 <= M <= 24000
// All a_j's >= 1
// {2, k, "k - 2 1's"} shows that 2k is a product-sum number

let elements = [];

for (let n1 = 2; n1 <= 24000; n1++) {
    for (let n2 = n1; n1 * n2 <= 24000; n2++) {
        elements.push([n1, n2]);
        for (let n3 = n2; n1 * n2 * n3 <= 24000; n3++) {
            elements.push([n1, n2, n3]);
            for (let n4 = n3; n1 * n2 * n3 * n4 <= 24000; n4++) {
                elements.push([n1, n2, n3, n4]);
                for (let n5 = n3; n1 * n2 * n3 * n4 * n5 <= 24000; n5++) {
                    elements.push([n1, n2, n3, n4, n5]);
                    for (let n6 = n5; n1 * n2 * n3 * n4 * n5 * n6 <= 24000; n6++) {
                        elements.push([n1, n2, n3, n4, n5, n6]);
                        for (let n7 = n6; n1 * n2 * n3 * n4 * n5 * n6 * n7 <= 24000; n7++) {
                            elements.push([n1, n2, n3, n4, n5, n6, n7]);
                            for (let n8 = n7; n1 * n2 * n3 * n4 * n5 * n6 * n7 * n8 <= 24000; n8++) {
                                elements.push([n1, n2, n3, n4, n5, n6, n7, n8]);
                                for (let n9 = n8; n1 * n2 * n3 * n4 * n5 * n6 * n7 * n8 * n9 <= 24000; n9++) {
                                    elements.push([n1, n2, n3, n4, n5, n6, n7, n8, n9]);
                                    for (let n10 = n9; n1 * n2 * n3 * n4 * n5 * n6 * n7 * n8 * n9 * n10 <= 24000; n10++) {
                                        elements.push([n1, n2, n3, n4, n5, n6, n7, n8, n9, n10]);
                                        for (let n11 = n10; n1 * n2 * n3 * n4 * n5 * n6 * n7 * n8 * n9 * n10 * n11 <= 24000; n11++) {
                                            elements.push([n1, n2, n3, n4, n5, n6, n7, n8, n9, n10, n11]);
                                            for (let n12 = n11; n1 * n2 * n3 * n4 * n5 * n6 * n7 * n8 * n9 * n10 * n11 * n12 <= 24000; n12++) {
                                                elements.push([n1, n2, n3, n4, n5, n6, n7, n8, n9, n10, n11, n12]);
                                                for (let n13 = n12; n1 * n2 * n3 * n4 * n5 * n6 * n7 * n8 * n9 * n10 * n11 * n12 * n13 <= 24000; n13++) {
                                                    elements.push([n1, n2, n3, n4, n5, n6, n7, n8, n9, n10, n11, n12, n13]);
                                                    for (let n14 = n13; n1 * n2 * n3 * n4 * n5 * n6 * n7 * n8 * n9 * n10 * n11 * n12 * n13 * n14 <= 24000; n14++) {
                                                        elements.push([n1, n2, n3, n4, n5, n6, n7, n8, n9, n10, n11, n12, n13, n14]);
                                                    }
                                                }
                                            }
                                        }
                                    }                         
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

const lengthAndProduct = (arr) => {
    const prod = arr.reduce((p,c) => p * c, 1);
    const sum = arr.reduce((p,c) => p + c, 0);
    const len = prod - sum + arr.length;
    return [len, prod];
}

let minimalPSN = Array(12001);
minimalPSN.fill(0);


for (const arr of elements) {
    const lp = lengthAndProduct(arr);
    const l = lp[0];
    const p = lp[1];
    if (l> 1 && l < 12001) {
        if (minimalPSN[l] === 0) {
            minimalPSN[l] = p;
        } else {
            minimalPSN[l] = Math.min(minimalPSN[l], p);
        }
        
    }
}


const uniquePSN = [... new Set(minimalPSN)];

const sum = uniquePSN.reduce((p,c) => p + c, 0);

const end = Date.now();

console.log("Execute time: ", end - start, " The sum of minimal product-sum numbers is: ", sum);